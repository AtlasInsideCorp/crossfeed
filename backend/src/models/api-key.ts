import {
  Entity,
  Index,
  Column,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  CreateDateColumn,
  BaseEntity,
  ManyToOne
} from 'typeorm';
import { Organization, User } from './';

@Entity()
export class ApiKey extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @ManyToOne((type) => User, (user) => user.apiKeys, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  user: User;

  @Column({
    type: 'timestamp',
    nullable: true
  })
  lastUsed: Date | null;

  @Column({
    type: 'text',
    nullable: true
  })
  key: string;
}
