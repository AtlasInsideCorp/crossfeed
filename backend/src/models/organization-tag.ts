import {
  Entity,
  Index,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  CreateDateColumn,
  BaseEntity,
  ManyToMany,
  JoinTable,
  Column
} from 'typeorm';
import { Organization, Scan } from '.';

@Entity()
@Index(['name'], { unique: true })
export class OrganizationTag extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @Column()
  name: string;

  @ManyToMany((type) => Organization, (organization) => organization.tags, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  @JoinTable()
  organizations: Organization[];

  @ManyToMany((type) => Scan, (scan) => scan.tags, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  @JoinTable()
  scans: Scan[];
}
