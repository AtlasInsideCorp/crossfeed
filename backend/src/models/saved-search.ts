import {
  Entity,
  Index,
  Column,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  CreateDateColumn,
  BaseEntity,
  OneToMany,
  ManyToMany,
  ManyToOne
} from 'typeorm';
import { Domain, Role, Scan, ScanTask } from '.';
import { User } from './user';

@Entity()
export class SavedSearch extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @Column()
  name: string;

  @Column()
  searchTerm: string;

  @Column()
  count: number;

  @Column({ type: 'jsonb', default: '[]' })
  filters: { field: string; values: any[]; type: string }[];

  @Column()
  searchPath: string;

  @ManyToOne((type) => User, {
    onDelete: 'SET NULL',
    onUpdate: 'CASCADE'
  })
  createdBy: User;
}
