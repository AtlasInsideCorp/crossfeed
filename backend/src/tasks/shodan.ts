import { Domain, Service, Vulnerability } from '../models';
import { plainToClass } from 'class-transformer';
import * as portscanner from 'portscanner';
import getIps from './helpers/getIps';
import { CommandOptions } from './ecs-client';
import saveServicesToDb from './helpers/saveServicesToDb';
import { chunk } from 'lodash';
import axios from 'axios';
import { IpToDomainsMap, sanitizeStringField } from './censysIpv4';
import saveVulnerabilitiesToDb from './helpers/saveVulnerabilitiesToDb';

// Shodan allows searching up to 100 IPs at once
const CHUNK_SIZE = 100;

interface ShodanResponse {
  ip_str: string;
  country_name: string;
  domains: string[];
  hostnames: string[];
  org: string;
  ports: number[];
  os: string;
  city: string;
  longitude: number;
  latitude: number;
  data: {
    port: number;
    product: string;
    data: string;
    cpe: string[];
    version: string;
    vulns: {
      [title: string]: {
        verified: boolean;
        references: string[];
        cvss: number;
        summary: string;
      };
    };
  }[];
}

export const handler = async (commandOptions: CommandOptions) => {
  const { organizationId, organizationName } = commandOptions;

  console.log('Running shodan on organization', organizationName);

  const domainsWithIPs = await getIps(organizationId);

  const ipToDomainsMap: IpToDomainsMap = domainsWithIPs.reduce<IpToDomainsMap>(
    (map: IpToDomainsMap, domain: Domain) => {
      if (!map[domain.ip]) {
        map[domain.ip] = [];
      }
      map[domain.ip].push(domain);
      return map;
    },
    {}
  );

  const chunks = chunk(domainsWithIPs, CHUNK_SIZE);

  for (const domainChunk of chunks) {
    const { data } = await axios.get<ShodanResponse[]>(
      `https://api.shodan.io/shodan/host/${domainChunk
        .map((domain) => domain.ip)
        .join(',')}?key=${process.env.SHODAN_API_KEY}`
    );
    const services: Service[] = [];
    const vulns: Vulnerability[] = [];
    for (const item of data) {
      const domains = ipToDomainsMap[item.ip_str];
      for (const domain of domains) {
        for (const service of item.data) {
          services.push(
            plainToClass(Service, {
              domain: domain,
              discoveredBy: { id: commandOptions.scanId },
              port: service.port,
              lastSeen: new Date(Date.now()),
              banner: sanitizeStringField(service.data),
              shodanResults: {
                product: service.product,
                version: service.version,
                cpe: service.cpe
              }
            })
          );
          if (service.vulns) {
            for (const cve in service.vulns) {
              vulns.push(
                plainToClass(Vulnerability, {
                  domain: domain,
                  lastSeen: new Date(Date.now()),
                  title: cve,
                  cve: cve,
                  cpe:
                    service.cpe && service.cpe.length > 0
                      ? service.cpe[0]
                      : null,
                  cvss: service.vulns[cve].cvss,
                  state: 'open',
                  source: 'shodan',
                  description: service.vulns[cve].summary,
                  needsPopulation: true
                })
              );
            }
          }
        }
      }
    }
    await saveServicesToDb(services);
    await saveVulnerabilitiesToDb(vulns, false);
  }

  console.log(`Shodan finished for ${domainsWithIPs.length} domains`);
};
