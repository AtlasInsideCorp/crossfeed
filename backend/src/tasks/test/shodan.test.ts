import { connectToDatabase, Domain, Organization, Scan } from '../../models';
import { handler as shodan } from '../shodan';
jest.mock('../helpers/getIps');
jest.mock('../helpers/saveServicesToDb');

const RealDate = Date;

describe('shodan', () => {
  let organization;
  let scan;
  beforeEach(async () => {
    await connectToDatabase();
    global.Date.now = jest.fn(() => new Date('2019-04-22T10:20:30Z').getTime());
    organization = await Organization.create({
      name: 'test-' + Math.random(),
      rootDomains: ['test-' + Math.random()],
      ipBlocks: [],
      isPassive: false
    }).save();
    scan = await Scan.create({
      name: 'shodan',
      arguments: {},
      frequency: 999
    }).save();
    await Domain.create({
      name: 'first_file_testdomain1',
      ip: '153.126.148.60',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain2',
      ip: '31.134.10.156',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain3',
      ip: '153.126.148.61',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain4',
      ip: '85.24.146.152',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain5',
      ip: '45.79.207.117',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain6',
      ip: '156.249.159.119',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain7',
      ip: '221.10.15.220',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain8',
      ip: '81.141.166.145',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain9',
      ip: '24.65.82.187',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain10',
      ip: '52.74.149.117',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain11',
      ip: '31.134.10.156',
      organization
    }).save();
    await Domain.create({
      name: 'first_file_testdomain12',
      ip: '1.1.1.1',
      organization
    }).save();
  });

  afterEach(async () => {
    global.Date = RealDate;
  });

  const checkDomains = async (organization) => {
    const domains = await Domain.find({
      where: { organization },
      relations: ['organization', 'services']
    });
    expect(
      domains
        .sort((a, b) => a.name.localeCompare(b.name))
        .map((e) => ({
          ...e,
          services: e.services.map((s) => ({
            ...s,
            id: null,
            updatedAt: null,
            createdAt: null
          })),
          organization: null,
          id: null,
          updatedAt: null,
          createdAt: null,
          syncedAt: null
        }))
    ).toMatchSnapshot();
    expect(domains.filter((e) => !e.organization).length).toEqual(0);
  };
  test('basic test', async () => {
    console.log(organization.id);
    await shodan({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: 'd0f51c16-a64a-4ed0-8373-d66485bfc678',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });
    await checkDomains(organization);
  });
});
