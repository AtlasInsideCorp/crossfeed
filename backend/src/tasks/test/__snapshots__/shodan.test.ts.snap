// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shodan basic test 1`] = `
Array [
  Object {
    "asn": null,
    "censysCertificatesResults": Object {},
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "153.126.148.60",
    "name": "first_file_testdomain1",
    "organization": null,
    "reverseName": "first_file_testdomain1",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "censysCertificatesResults": Object {},
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "1.1.1.1",
    "name": "first_file_testdomain12",
    "organization": null,
    "reverseName": "first_file_testdomain12",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "censysCertificatesResults": Object {},
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "31.134.10.156",
    "name": "first_file_testdomain2",
    "organization": null,
    "reverseName": "first_file_testdomain2",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
]
`;

exports[`shodan basic test: helpers.saveServicesToDb 1`] = `
Array [
  Service {
    "banner": "* OK [CAPABILITY IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE AUTH=PLAIN AUTH=LOGIN] Dovecot ready.
* CAPABILITY IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE AUTH=PLAIN AUTH=LOGIN
A001 OK Capability completed.
* ID NIL
A002 OK ID completed.
A003 BAD Error in IMAP command received by server.
",
    "discoveredBy": Object {
      "id": "d0f51c16-a64a-4ed0-8373-d66485bfc678",
    },
    "domain": Domain {
      "asn": null,
      "censysCertificatesResults": Object {},
      "cloudHosted": false,
      "country": null,
      "createdAt": 2020-12-22T04:16:59.642Z,
      "fromRootDomain": null,
      "id": "5abb8989-4f15-4cf9-b661-ead0d66e9796",
      "ip": "153.126.148.60",
      "name": "first_file_testdomain1",
      "organization": Organization {
        "createdAt": 2020-12-22T04:16:59.612Z,
        "id": "0071d91f-6e8a-4159-b958-7895b48a68a1",
        "inviteOnly": true,
        "ipBlocks": Array [],
        "isPassive": false,
        "name": "test-0.5888505160814219",
        "rootDomains": Array [
          "test-0.14487301045697154",
        ],
        "updatedAt": 2020-12-22T04:16:59.612Z,
      },
      "reverseName": "first_file_testdomain1",
      "screenshot": null,
      "ssl": null,
      "syncedAt": null,
      "updatedAt": 2020-12-22T04:16:59.642Z,
    },
    "lastSeen": 2019-04-22T10:20:30.000Z,
    "port": 993,
    "shodanResults": Object {
      "cpe": undefined,
      "product": undefined,
      "version": undefined,
    },
  },
]
`;

exports[`shodan basic test: helpers.saveServicesToDb 2`] = `
Array [
  Service {
    "banner": "
Recursion: enabled
Resolver ID: AMS",
    "discoveredBy": Object {
      "id": "d0f51c16-a64a-4ed0-8373-d66485bfc678",
    },
    "domain": Domain {
      "asn": null,
      "censysCertificatesResults": Object {},
      "cloudHosted": false,
      "country": null,
      "createdAt": 2020-12-22T04:16:59.663Z,
      "fromRootDomain": null,
      "id": "c5e5921d-a799-4f68-a441-311d8494e577",
      "ip": "1.1.1.1",
      "name": "first_file_testdomain12",
      "organization": Organization {
        "createdAt": 2020-12-22T04:16:59.612Z,
        "id": "0071d91f-6e8a-4159-b958-7895b48a68a1",
        "inviteOnly": true,
        "ipBlocks": Array [],
        "isPassive": false,
        "name": "test-0.5888505160814219",
        "rootDomains": Array [
          "test-0.14487301045697154",
        ],
        "updatedAt": 2020-12-22T04:16:59.612Z,
      },
      "reverseName": "first_file_testdomain12",
      "screenshot": null,
      "ssl": null,
      "syncedAt": null,
      "updatedAt": 2020-12-22T04:16:59.663Z,
    },
    "lastSeen": 2019-04-22T10:20:30.000Z,
    "port": 53,
    "shodanResults": Object {
      "cpe": undefined,
      "product": undefined,
      "version": undefined,
    },
  },
]
`;
